### PR TITLE
Evo2 Mamba

### PR DESCRIPTION
**Pull request by jstjohn**
### Type of changes

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Refactor
- [ ]  Documentation update
- [ ]  Other (please describe):

### CI Pipeline Configuration
Configure CI behavior by applying the relevant labels:

- [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest
- [INCLUDE_SLOW_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_slow_tests) - Execute tests labelled as slow in pytest for extensive testing

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

* If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
* If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Usage

```python
TODO: Add code snippet
```

### Pre-submit Checklist

 - [ ] I have tested these changes locally
 - [ ] I have updated the documentation accordingly
 - [ ] I have added/updated tests as needed
 - [ ] All existing tests pass successfully

### CODE REVIEW

**Review by dorotat-nv — DISMISSED** [review]
Before merging it, lets make sure JET config for partial-conv is added. Please reach out to me for guidance if needed

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py [hunk 1]
```diff
@@ -453,3 +491,112 @@ def main():
...
+                micro_batch_idx=0,
+                sequence_parallel=False,
+                micro_batch_size_per_context_rank=None,
+                context_parallel_size=self.trainer.strategy._cp_size,
+                sequence_length=self.config.seq_length,
+                multiple_of=1,
+            )
+            inference_data["attention_mask"] = None
+
+        with torch.no_grad():
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py]: We probably need to fix this here like we did for evo2, that issue with our inference_context and no_grad.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/utils/loss/embedding_variance.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+# from megatron.core import parallel_state # Example import
+
+
+import torch
+import torch.distributed
+from megatron.core import parallel_state
+from torch.autograd import Function
+
+
+class SquaredErrorTargetedVarianceLossFunction(Function):
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/utils/loss/embedding_variance.py:27]: TODO simplify this code. This likely has a lot of cursor auto-generated cruft.

File: Dockerfile [hunk 3]
```diff
@@ -135,8 +135,19 @@ fi
 RUN apt-get update -qy && apt-get install -y libopenblas-dev && pip install scikit-misc==0.3.1
 
 # Mamba dependancy installation
-RUN pip --disable-pip-version-check --no-cache-dir install \
-  git+https://github.com/state-spaces/mamba.git@v2.2.2 --no-deps
+# See https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/blob/d3c853c2d/docker/Dockerfile#L193-198
+#  for the command we want to keep in sync. Note that the package install is modified slightly to first build a
+#  wheel (no deps) and then install the wheel. This avoids issues with torch version conflicts.
+RUN git clone https://github.com/state-spaces/mamba.git && \
```
Comment (pstjohn) [inline Dockerfile]: can you `rm -f mamba/` as well

cleanest solution would be to have a separate build stage where you make the wheel and then copy it over, but 🤷

**Comment by jwilber** [comment]
> Before merging it, lets make sure JET config for partial-conv is added. Please reach out to me for guidance if needed

@dorotat-nv @jstjohn 
I don't think we can get to this before the merge, but I have a related epic in August during which point I can add this in.

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/test_evo2_mamba_batch_generate.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        ckpt_dir: Path = load(ckpt_dir_or_name)
+    from nemo.collections.llm import inference
+
+    inference_wrapped_model, mcore_tokenizer = inference.setup_model_and_tokenizer(
+        path=ckpt_dir,
+        trainer=trainer,
+        params_dtype=torch.bfloat16,
+        inference_batch_times_seqlen_threshold=8192,  # TODO
+        inference_max_seq_length=8192,  # TODO
+        # use_te_rng_tracker=True,
```
Comment (jwilber) [inline sub-packages/bionemo-evo2/tests/bionemo/evo2/test_evo2_mamba_batch_generate.py]: nit: we can probably remove these commented args so as to not confuse the bcr folks (if they copy this test over)

**Review by tshimko-nv — APPROVED** [review]

**Comment by trvachov** [comment]
@jstjohn I have dismissed Dorota's review under the assumption her comments were addressed. (She is OOO, cc @dorotat-nv ).

This should unblock the usual merge process.

**Comment by trvachov** [comment]
looks like failing slow tests + notebooks...
